rfcs/0000-unified-configuration.ftd

```diff
@@ -0,0 +1,312 @@
+
+Add a virtual module (`<pkg-name>/-/config`) similar to existing
+[assets](https://fastn.com/assets) module. 
+
+This module will allow access to variables defined in `FASTN.ftd` file.
+
+-- ds.code: Example
+lang: ftd
+
+\-- import: my-package/-/config \;; This is new!
+
+\-- import: fastn/processors as pr
+
+\-- person list people:
+$processor$: pr.pg
+use: $config.my-db \;; defined in my-package's FASTN.ftd
+
+SELECT * FROM users;
+
+-- ds.h2: Configuring dependencies and apps
+
+`FASTN.ftd` file will be central to the whole configuration system. Users can
+modify the behaviour of a dependency by passing variables when they're added.
+Example:
+
+-- ds.code: `FASTN.ftd` of test-todos
+lang: ftd
+
+\-- import: fastn
+
+\-- fastn.package: test-todos
+
+\;; fastn.app works in similar way
+\-- fastn.dependency: amitu.com/todos
+max-todo-per-page: 30
+
+-- end: ds.code
+
+-- ds.markdown:
+
+The above snippet is configuring `amitu.com/todos` by modifying variables that
+are in its `FASTN.ftd` file.
+
+The `FASTN.ftd` of `amitu.com/todos` will look like this:
+
+-- ds.code: FASTN.ftd of amitu.com/todos
+lang: ftd
+
+\-- import: fastn
+
+\-- fastn.package: amitu.com/todos
+
+\-- integer max-todo-per-page: 10 \;; becomes 30 when used by test-todos
+
+\-- string sample-title: TODO App \;; this was not modified by test-todos so this stays intact
+
+-- end: ds.code
+
+-- ds.markdown:
+
+- While adding dependency, users can modify any variable defined in the
+  `FASTN.ftd` file of the said dependency.
+
+- Any extra header value passed to `fastn.dependency` will result in an error.
+
+-- ds.h2: Optional vs. Required dependency configuration
+
+A package can create variables in `FASTN.ftd` in different ways which will
+dictate if they're required or optional. See the example below:
+
+-- ds.code: `FASTN.ftd` of `amitu.com/todos`
+lang: ftd
+
+\-- import: fastn
+
+\-- fastn.package: amitu.com/todos
+db: $my-db-for-dev \;; this value for db is available only when it's not used as a dependency
+
+\-- fastn.db-conn db:  \;; this has no value so this will be required
+
+\-- integer max-todos: 30 \;; this has a default value so this is optional
+
+-- ds.markdown:
+
+- `db` must be defined when the package is being used, and fastn.package is one
+  such usage! This allows for easy testing, say if I am building
+  amitu.com/todos, I do not have to create a test app, and I can work directly
+  in amitu.com/todos package, and my test values would be passed there.
+
+- `max-todos` is defined with a value so this doesn't have to be passed as a
+  header when using `amitu.com/todos` as a dependency.
+
+-- ds.code: Example use of `amitu.com/todos` in `my-website`
+lang: ftd
+
+\-- import: fastn
+
+\-- fastn.package: my-website
+
+\-- fastn.dependency: amitu.com/todos
+db: $my-db \;; this is required otherwise an error will be thrown
+\;; max-todos is optional so we ignore this
+
+
+-- ds.h3: Q. Why do we allow defining config variables outside of `FASTN.ftd`?
```

Comment:
This should be out of scope of this RFC. I am not sure if we should allow this (as I am leaning towards making `fastn.env.*` only accessible from FASTN.ftd, and defining config variable outside FASTN.ftd without env would be tricky / unsafe (people will hardcode values).

Reply:
Okay. I'll change it to only say that we allow defining config variables in FASTN.ftd for now